ops/testing.py

```diff
@@ -2931,23 +2931,49 @@ def add_layer(
         else:
             self._layers[label] = layer_obj
 
-    def _render_services(self) -> Dict[str, pebble.Service]:
-        services: Dict[str, pebble.Service] = {}
+    def _render_services(self) -> Dict[str, pebble.ServiceDict]:
+        services: Dict[str, pebble.ServiceDict] = {}
         for key in sorted(self._layers.keys()):
             layer = self._layers[key]
             for name, service in layer.services.items():
                 # TODO: (jam) 2021-04-07 have a way to merge existing services
-                services[name] = service
+                services[name] = service.to_dict()
         return services
 
+    def _render_checks(self) -> Dict[str, pebble.CheckDict]:
+        services: Dict[str, pebble.CheckDict] = {}
+        for key in sorted(self._layers.keys()):
+            layer = self._layers[key]
+            for name, check in layer.checks.items():
+                services[name] = check.to_dict()
+        return services
+
+    def _render_log_targets(self) -> Dict[str, pebble.LogTargetDict]:
+        services: Dict[str, pebble.LogTargetDict] = {}
+        for key in sorted(self._layers.keys()):
+            layer = self._layers[key]
+            for name, log_target in layer.log_targets.items():
+                services[name] = log_target.to_dict()
+        return services
+
+    def _render_plan(self) -> str:
+        raw_plan = {}
+        services = self._render_services()
+        if services:
+            raw_plan["services"] = services
+
+        checks = self._render_checks()
+        if checks:
+            raw_plan["checks"] = checks
+
+        log_targets = self._render_log_targets()
+        if log_targets:
+            raw_plan["log-targets"] = log_targets
+        return yaml.safe_dump(raw_plan)
+
     def get_plan(self) -> pebble.Plan:
         self._check_connection()
-        plan = pebble.Plan('{}')
-        services = self._render_services()
-        if not services:
-            return plan
-        for name in sorted(services.keys()):
-            plan.services[name] = services[name]
+        plan = pebble.Plan(self._render_plan())
```

Comment:
I don't love the way we're rendering to a YAML string, only to load it back from YAML to objects again. Can we do this more similarly to how it was?

Reply:
I have a deep feeling of unease towards mutation that was hammered into me at a young age, so I'll roll this back but add a fixme.